File: qmdc-py/qmdc/parsers/field.py
"""Field parser - extracts fields from list items."""

import re
from typing import Any

from markdown_it.token import Token

# Pre-compiled regexes for field parsing (compiled once at module load)
_FIELD_PATTERN = re.compile(r"^([a-zA-Z_][a-zA-Z0-9_]*)\s*:\s*(.*)$", re.DOTALL)
_INVALID_FIELD_LIKE_PATTERN = re.compile(r"^([^:]+):\s+(.*)$", re.DOTALL)
_VALID_KEY_PATTERN = re.compile(r"^[a-zA-Z_][a-zA-Z0-9_]*$")


def parse_yaml_array(value_str: str) -> tuple[list[Any], dict[str, str]]:
    """
    Parse YAML-style array like [a, b, c] or [1, 2, 3].

    Returns: (items_list, item_types)
    """
    # Remove brackets
    inner = value_str[1:-1].strip()

    if not inner:
        return [], {}

    items: list[Any] = []
    types: dict[str, str] = {}

    # Split by comma, handle quoted strings
    parts = _split_yaml_array(inner)

    for idx, part in enumerate(parts):
        val, type_name = parse_field_value(part)
        items.append(val)
        types[str(idx)] = type_name

    return items, types


def _split_yaml_array(s: str) -> list[str]:
    """Split YAML array content by commas, respecting quotes."""
    result: list[str] = []
    current = ""
    in_quotes = False
    quote_char = ""

    for char in s:
        if char in ('"', "'") and not in_quotes:
            in_quotes = True
            quote_char = char
            current += char
        elif char == quote_char and in_quotes:
            in_quotes = False
            current += char
            quote_char = ""
        elif char == "," and not in_quotes:
            result.append(current.strip())
            current = ""
        else:
            current += char

    if current.strip():
        result.append(current.strip())

    return result


def parse_field_value(value_str: str) -> tuple[Any, str]:
    """
    Parse field value and auto-detect type.

    Returns: (value, type_name)

    Types:
    - array → (list, "array")
    - null → (None, "null")
    - true/false → (bool, "boolean")
    - number → (int/float, "number")
    - string → (str, "string")
    """
    value = value_str.strip()

    # Empty array
    if value == "[]":
        return [], "array"

    # Multiple comma-separated references: [[#a]], [[#b]], [[#c]]
    # NOT a YAML array (which would be [[[#a]], [[#b]]])
    if value.startswith("[[") and not value.startswith("[[[") and "]], [[" in value:
        items = []
        for part in value.split("]], [["):
            part = part.strip()
            # Restore brackets
            if not part.startswith("[["):
                part = "[[" + part
            if not part.endswith("]]"):
                part = part + "]]"
            items.append(part)
        return items, "ref_array"

    # YAML array [a, b, c] - but not single references [[#id]]
    # Array of refs [[[#id1]], [[#id2]]] should be parsed as array
    is_array_bracket = value.startswith("[") and value.endswith("]")
    is_single_ref = value.startswith("[[") and not value.startswith("[[[")
    if is_array_bracket and not is_single_ref:
        items, _ = parse_yaml_array(value)
        return items, "array"

    # null
    if value == "null":
        return None, "null"

    # boolean
    if value == "true":
        return True, "boolean"
    if value == "false":
        return False, "boolean"

    # number (int or float)
    try:
        # Try int first
        if "." not in value:
            return int(value), "number"
        # Then float
        return float(value), "number"
    except ValueError:
        pass

    # string (default) - remove quotes if present
    if (value.startswith('"') and value.endswith('"')) or (
        value.startswith("'") and value.endswith("'")
    ):
        return value[1:-1], "string"

    return value, "string"


def parse_fields_from_list(
    tokens: list[Token], start_idx: int, block_tree: Any = None, *, raw_strings: bool = False
) -> tuple[
    dict[str, Any], dict[str, str], dict[str, str], list[dict[str, Any]], int, list[dict[str, Any]]
]:
    """
    Parse fields from markdown list starting at start_idx.

    Returns:
        (fields_dict, types_dict, syntax_dict, invalid_items, next_index, nested_subitems_errors)

    invalid_items: list of {"key": str, "content": str, "line": int, "after": str} for items
                   that look like fields but have invalid keys (e.g. Cyrillic).
                   "after" is the last valid field key before this item, or "__self".
    nested_subitems_errors: list of {"key": str, "line": int} for fields with nested sub-items
                           (pattern `- key:\n  - item` which is forbidden).
    """
    fields: dict[str, Any] = {}
    types: dict[str, str] = {}
    syntax: dict[str, str] = {}
    invalid_items: list[dict[str, Any]] = []
    nested_subitems_errors: list[dict[str, Any]] = []
    i = start_idx

    # Pattern: `- key: value` or `- key:value` (with DOTALL for multiline values)
    field_pattern = _FIELD_PATTERN
    # Pattern for invalid field-like items (any text with colon and space)
    invalid_field_like_pattern = _INVALID_FIELD_LIKE_PATTERN
    # Valid key pattern
    valid_key_pattern = _VALID_KEY_PATTERN

    current_line = 0
    last_valid_field: str = "__self"

    while i < len(tokens):
        token = tokens[i]

        if token.type == "bullet_list_open":
            i += 1
            continue

        if token.type == "bullet_list_close":
            i += 1
            break

        if token.type == "list_item_open":
            # Track line number for error reporting
            if token.map:
                current_line = token.map[0] + 1  # 1-based
            i += 1
            continue

        if token.type == "list_item_close":
            i += 1
            continue

        if token.type == "paragraph_open":
            i += 1
            continue

        if token.type == "paragraph_close":
            i += 1
            continue

        if token.type == "inline":
            # Parse field from inline content
            content = token.content.strip()
            # Get line from token map if available
            if token.map:
                current_line = token.map[0] + 1
            first_line = content.split("\n")[0]
            match = field_pattern.match(first_line)

            if match:
                key = match.group(1)
                # For multiline content, get value from full content
                colon_pos = content.find(":")
                value_str = content[colon_pos + 1 :].strip() if colon_pos >= 0 else match.group(2)

                # Check for YAML multiline: `field: |`
                # markdown-it may merge `field: |` with next line into one inline token,
                # so value_str can be `"|"` or `"|\n  next line..."`.
                if value_str.strip() == "|" or value_str.strip().startswith("|\n"):
                    # Use raw-slice from BlockTree when available (preserves numbering, formatting)
                    if block_tree and token.map:
                        # Find the OUTER list_item_close (track nesting to skip inner ones)
                        scan = i + 1
                        nesting = 0
                        while scan < len(tokens):
                            st = tokens[scan].type
                            if st in ("ordered_list_open", "bullet_list_open"):
                                nesting += 1
                            elif st in ("ordered_list_close", "bullet_list_close"):
                                nesting -= 1
                            elif st == "list_item_close" and nesting == 0:
                                break
                            scan += 1
                        pipe_line = token.map[0]  # line of "- field: |"
                        # End line: next list_item_open or bullet_list_close
                        end_line = None
                        for j in range(scan, min(scan + 3, len(tokens))):
                            if j < len(tokens) and tokens[j].type == "list_item_open":
                                if tokens[j].map:
                                    end_line = tokens[j].map[0]
                                break
                            if j < len(tokens) and tokens[j].type == "bullet_list_close":
                                break
                        if end_line is None:
                            # Last item — find bullet_list_close
                            for j in range(scan, min(scan + 3, len(tokens))):
                                if j < len(tokens) and tokens[j].type == "bullet_list_close":
                                    if tokens[j].map:
                                        end_line = tokens[j].map[0]
                                    break
                        if end_line is None:
                            end_line = block_tree.line_count

                        # Extract raw content after the pipe line
                        raw = block_tree.get_lines_raw(pipe_line + 1, end_line)
                        # Strip trailing blank lines
                        raw_stripped = raw.rstrip("\n")
                        # Dedent: remove common leading whitespace
                        raw_lines = raw_stripped.split("\n")
                        if raw_lines:
                            indents = [len(ln) - len(ln.lstrip()) for ln in raw_lines if ln.strip()]
                            min_indent = min(indents) if indents else 0
                            raw_lines = [ln[min_indent:] for ln in raw_lines]
                        value = "\n".join(raw_lines)

                        fields[key] = value
                        types[key] = "string"
                        syntax[key] = "yaml_multiline"
                        last_valid_field = key
                        # Skip to list_item_close
                        i = scan
                        continue
                    else:
                        # Fallback: collect from tokens
                        multiline_parts: list[str] = []
                        if value_str.strip().startswith("|\n"):
                            after_pipe = value_str.strip()[2:]
                            lines = after_pipe.split("\n")
                            if lines:
                                indents = [len(ln) - len(ln.lstrip()) for ln in lines if ln.strip()]
                                min_indent = min(indents) if indents else 0
                                lines = [
                                    ln[min_indent:] if len(ln) >= min_indent else ln for ln in lines
                                ]
                            multiline_parts.append("\n".join(lines))

                        i += 1
                        nesting = 0
                        while i < len(tokens) and tokens[i].type != "list_item_close":
                            t = tokens[i]
                            if t.type == "fence":
                                lang = t.info or ""
                                fence_content = t.content.rstrip("\n")
                                if lang:
                                    multiline_parts.append(f"```{lang}\n{fence_content}\n```")
                                else:
                                    multiline_parts.append(f"```\n{fence_content}\n```")
                            elif t.type == "code_block":
                                multiline_parts.append(t.content.rstrip("\n"))
                            elif t.type == "inline":
                                multiline_parts.append(t.content)
                            elif t.type in (
                                "ordered_list_open",
                                "bullet_list_open",
                            ):
                                nesting += 1
                            elif t.type in (
                                "ordered_list_close",
                                "bullet_list_close",
                            ):
                                nesting -= 1
                            elif t.type in (
                                "paragraph_open",
                                "paragraph_close",
                                "list_item_open",
                                "list_item_close",
                            ):
                                pass
                            i += 1

                        value = "\n".join(multiline_parts) if multiline_parts else ""
                        fields[key] = value
                        types[key] = "string"
                        syntax[key] = "yaml_multiline"
                        last_valid_field = key
                        continue

                if raw_strings:
                    fields[key] = value_str
                    types[key] = "string"
                    last_valid_field = key
                else:
                    value, type_name = parse_field_value(value_str)
                    fields[key] = value
                    types[key] = "array" if type_name == "ref_array" else type_name
                    last_valid_field = key

                    # Track syntax for arrays
                    if type_name == "ref_array":
                        syntax[key] = "comma_refs"
                    elif type_name == "array":
                        # Detect multiline array: value spans multiple lines
                        if "\n" in value_str and value_str.strip().startswith("["):
                            syntax[key] = "yaml_multiline_array"
                        else:
                            syntax[key] = "yaml_array"

                # Detect nested sub-items: field with empty value followed by nested list
                # e.g. `- affected_files:\n  - item1\n  - item2`
                # This is a syntax error (nested_subitems) per spec rule_no_nested_subitems
                if value_str == "" and i + 1 < len(tokens):
                    # Look ahead for nested list (bullet or ordered)
                    # before list_item_close
                    lookahead = i + 1
                    while lookahead < len(tokens) and tokens[lookahead].type in (
                        "paragraph_close",
                    ):
                        lookahead += 1
                    if lookahead < len(tokens) and tokens[lookahead].type in (
                        "bullet_list_open",
                        "ordered_list_open",
                    ):
                        nested_list_type = tokens[lookahead].type
                        nested_list_close = nested_list_type.replace("_open", "_close")
                        # Skip past the nested list
                        lookahead += 1  # skip list_open
                        while (
                            lookahead < len(tokens) and tokens[lookahead].type != nested_list_close
                        ):
                            lookahead += 1
                        if lookahead < len(tokens):
                            lookahead += 1  # skip list_close
                        # Record as nested_subitems error
                        nested_subitems_errors.append(
                            {
                                "key": key,
                                "line": current_line,
                            }
                        )
                        # Remove the field we just added (it has empty value)
                        del fields[key]
                        if key in types:
                            del types[key]
                        i = lookahead
                        continue
            else:
                # Not a valid field - check if it looks like a field with invalid key
                invalid_match = invalid_field_like_pattern.match(first_line)
                if invalid_match:
                    potential_key = invalid_match.group(1).strip()
                    if potential_key and not valid_key_pattern.match(potential_key):
                        # Invalid field key (e.g. Cyrillic)
                        invalid_items.append(
                            {
                                "key": potential_key,
                                "content": f"- {content}",
                                "line": current_line,
                                "after": last_valid_field,
                            }
                        )
                elif last_valid_field != "__self" and content:
                    # Plain list item after a valid field — not a field at all.
                    # Save as invalid item so it can be preserved in __comments.
                    invalid_items.append(
                        {
                            "key": "",
                            "content": f"- {content}",
                            "line": current_line,
                            "after": last_valid_field,
                        }
                    )

            i += 1
            continue

        # Unknown token, skip
        i += 1

    return fields, types, syntax, invalid_items, i, nested_subitems_errors


def parse_array_items_from_list(tokens: list[Token], start_idx: int) -> tuple[list[Any], int]:
    """
    Parse list items as array elements (no key: prefix).

    Used for [[field: array]] sections where list items are plain values.
    Only bullet lists reach this function — ordered lists are intercepted
    by the parser and emitted as ordered_list_in_array errors.

    Returns:
        (items_list, next_index)
    """
    items: list[Any] = []
    i = start_idx
    nesting = 0

    while i < len(tokens):
        token = tokens[i]
        if token.type in ("bullet_list_open", "ordered_list_open"):
            nesting += 1
            i += 1
            continue
        if token.type in ("bullet_list_close", "ordered_list_close"):
            nesting -= 1
            if nesting <= 0:
                i += 1
                break
            i += 1
            continue
        if token.type == "list_item_open":
            i += 1
            continue
        if token.type == "list_item_close":
            i += 1
            continue
        if token.type == "paragraph_open":
            i += 1
            continue
        if token.type == "paragraph_close":
            i += 1
            continue
        if token.type == "inline":
            content = token.content.strip()
            value, _ = parse_field_value(content)
            items.append(value)
            i += 1
            continue
        i += 1

    return items, i

File: qmdc-ts/src/parsers/field.ts
/**
 * Field parser - extracts fields from list items
 */

import type Token from 'markdown-it/lib/token';
import type { BlockTree } from '../block_tree.js';

/**
 * Parse YAML-style array like [a, b, c] or [1, 2, 3]
 *
 * Returns: [items_list, item_types, item_raw_tokens]
 * item_raw_tokens[idx] holds the raw source token for integer-valued float
 * literals (e.g. "1.0", "2.000") that JS collapses to an int; undefined for
 * every other element. Used to restore canonical `X.0` form in the JSON data
 * column without changing the parsed numeric value.
 */
function parseYamlArray(
  valueStr: string
): [unknown[], Record<string, string>, (string | undefined)[]] {
  // Remove brackets
  const inner = valueStr.slice(1, -1).trim();

  if (!inner) {
    return [[], {}, []];
  }

  const items: unknown[] = [];
  const types: Record<string, string> = {};
  const rawTokens: (string | undefined)[] = [];

  // Split by comma, handle quoted strings
  const parts = splitYamlArray(inner);

  for (let idx = 0; idx < parts.length; idx++) {
    const part = parts[idx];
    if (part !== undefined) {
      const [val, typeName, rawTok] = parseFieldValue(part);
      items.push(val);
      types[String(idx)] = typeName;
      rawTokens.push(rawTok);
    }
  }

  return [items, types, rawTokens];
}

/**
 * Split YAML array content by commas, respecting quotes
 */
function splitYamlArray(s: string): string[] {
  const result: string[] = [];
  let current = '';
  let inQuotes = false;
  let quoteChar = '';

  for (const char of s) {
    if ((char === '"' || char === "'") && !inQuotes) {
      inQuotes = true;
      quoteChar = char;
      current += char;
    } else if (char === quoteChar && inQuotes) {
      inQuotes = false;
      current += char;
      quoteChar = '';
    } else if (char === ',' && !inQuotes) {
      result.push(current.trim());
      current = '';
    } else {
      current += char;
    }
  }

  if (current.trim()) {
    result.push(current.trim());
  }

  return result;
}

/**
 * Parse field value and auto-detect type
 *
 * Returns: [value, type_name]
 *
 * Types:
 * - array → [array, "array"]
 * - null → [null, "null"]
 * - true/false → [boolean, "boolean"]
 * - number → [number, "number"]
 * - string → [string, "string"]
 */
export function parseFieldValue(
  valueStr: string
): [unknown, string, string | undefined, (string | undefined)[]?] {
  const value = valueStr.trim();

  // Empty array
  if (value === '[]') {
    return [[], 'array', undefined];
  }

  // Multiple comma-separated references: [[#a]], [[#b]], [[#c]]
  // NOT a YAML array (which would be [[[#a]], [[#b]]])
  if (value.startsWith('[[') && !value.startsWith('[[[') && value.includes(']], [[')) {
    const items: string[] = [];
    for (let part of value.split(']], [[')) {
      part = part.trim();
      // Restore brackets
      if (!part.startsWith('[[')) {
        part = '[[' + part;
      }
      if (!part.endsWith(']]')) {
        part = part + ']]';
      }
      items.push(part);
    }
    return [items, 'ref_array', undefined];
  }

  // YAML array [a, b, c] - but not single references [[#id]]
  // Array of refs [[[#id1]], [[#id2]]] should be parsed as array
  const isArrayBracket = value.startsWith('[') && value.endsWith(']');
  const isSingleRef = value.startsWith('[[') && !value.startsWith('[[[');
  if (isArrayBracket && !isSingleRef) {
    const [items, , itemRawTokens] = parseYamlArray(value);
    // Surface per-element raw float tokens only when at least one is present,
    // so callers can restore canonical `X.0` form in the JSON data column.
    const hasRaw = itemRawTokens.some((t) => t !== undefined);
    return [items, 'array', undefined, hasRaw ? itemRawTokens : undefined];
  }

  // null
  if (value === 'null') {
    return [null, 'null', undefined];
  }

  // boolean
  if (value === 'true') {
    return [true, 'boolean', undefined];
  }
  if (value === 'false') {
    return [false, 'boolean', undefined];
  }

  // number (int or float)
  if (/^-?\d+(\.\d+)?$/.test(value)) {
    const num = parseFloat(value);
    const raw = value.includes('.') && Number.isInteger(num) ? value : undefined;
    return [num, 'number', raw];
  }

  // string (default) - remove quotes if present
  if (
    (value.startsWith('"') && value.endsWith('"')) ||
    (value.startsWith("'") && value.endsWith("'"))
  ) {
    return [value.slice(1, -1), 'string', undefined];
  }

  return [value, 'string', undefined];
}

export interface InvalidFieldItem {
  key: string;
  content: string;
  line: number;
  after: string;
}

export interface NestedSubitemsError {
  key: string;
  line: number;
}

/**
 * Parse fields from markdown list starting at start_idx
 *
 * Returns: [fields_dict, types_dict, syntax_dict, invalid_items, next_index]
 */
export function parseFieldsFromList(
  tokens: Token[],
  startIdx: number,
  blockTree?: BlockTree,
  options?: { rawStrings?: boolean }
): [
  Record<string, unknown>,
  Record<string, string>,
  Record<string, string>,
  InvalidFieldItem[],
  number,
  Record<string, string>,
  NestedSubitemsError[],
] {
  const fields: Record<string, unknown> = {};
  const types: Record<string, string> = {};
  const syntax: Record<string, string> = {};
  const invalidItems: InvalidFieldItem[] = [];
  const nestedSubitemsErrors: NestedSubitemsError[] = [];
  const rawValues: Record<string, string> = {};
  let i = startIdx;

  // Pattern: `- key: value` or `- key:value`
  const fieldPattern = /^([a-zA-Z_][a-zA-Z0-9_]*)\s*:\s*(.*)$/;
  // Pattern for invalid field-like items (any text with colon and space)
  const invalidFieldLikePattern = /^([^:]+):\s+(.*)$/;
  // Valid key pattern
  const validKeyPattern = /^[a-zA-Z_][a-zA-Z0-9_]*$/;

  let lastValidField = '__self';

  while (i < tokens.length) {
    const token = tokens[i];
    if (!token) break;

    if (token.type === 'bullet_list_open') {
      i++;
      continue;
    }

    if (token.type === 'bullet_list_close') {
      i++;
      break;
    }

    if (token.type === 'list_item_open') {
      i++;
      continue;
    }

    if (token.type === 'list_item_close') {
      i++;
      continue;
    }

    if (token.type === 'paragraph_open') {
      i++;
      continue;
    }

    if (token.type === 'paragraph_close') {
      i++;
      continue;
    }

    if (token.type === 'inline') {
      // Parse field from inline content
      const content = token.content?.trim() ?? '';

      // Handle multiline content: markdown-it may include \n in inline tokens
      const firstLine = content.split('\n')[0] ?? '';
      const match = firstLine.match(fieldPattern);

      if (match) {
        const key = match[1];
        let valueStr = match[2] ?? '';
        if (key) {
          // Check for multiline YAML array: value starts with [ but doesn't close on same line
          let isMultilineArray = false;
          if (
            valueStr.trimStart().startsWith('[') &&
            !valueStr.trimEnd().endsWith(']') &&
            content.includes('\n')
          ) {
            isMultilineArray = true;
            // Join all lines, strip leading whitespace from continuation lines
            const joined = content
              .split('\n')
              .map((l) => l.trim())
              .join(' ');
            const fullMatch = joined.match(fieldPattern);
            if (fullMatch && fullMatch[2]) {
              valueStr = fullMatch[2].trim();
            }
          }

          // Check for YAML multiline: `field: |`
          // markdown-it may merge `field: |` with next line into one inline token,
          // so valueStr can be `"|"` or `"|\n  next line..."`.
          if (valueStr.trim() === '|' || valueStr.trim().startsWith('|\n')) {
            // Use raw-slice from BlockTree when available (preserves numbering, formatting)
            if (blockTree && token.map) {
              // Find the OUTER list_item_close (track nesting to skip inner ones)
              let scan = i + 1;
              let nesting = 0;
              while (scan < tokens.length) {
                const st = tokens[scan]?.type;
                if (st === 'ordered_list_open' || st === 'bullet_list_open') {
                  nesting++;
                } else if (st === 'ordered_list_close' || st === 'bullet_list_close') {
                  nesting--;
                } else if (st === 'list_item_close' && nesting === 0) {
                  break;
                }
                scan++;
              }
              const pipeLine = token.map[0]; // line of "- field: |"
              // End line: next list_item_open or bullet_list_close
              let endLine: number | undefined;
              for (let j = scan; j < Math.min(scan + 3, tokens.length); j++) {
                if (tokens[j]?.type === 'list_item_open') {
                  if (tokens[j]?.map) {
                    endLine = tokens[j]!.map![0];
                  }
                  break;
                }
                if (tokens[j]?.type === 'bullet_list_close') {
                  break;
                }
              }
              if (endLine === undefined) {
                // Last item — find bullet_list_close
                for (let j = scan; j < Math.min(scan + 3, tokens.length); j++) {
                  if (tokens[j]?.type === 'bullet_list_close') {
                    if (tokens[j]?.map) {
                      endLine = tokens[j]!.map![0];
                    }
                    break;
                  }
                }
              }
              if (endLine === undefined) {
                endLine = blockTree.lineCount;
              }

              // Extract raw content after the pipe line
              const raw = blockTree.getLinesRaw(pipeLine + 1, endLine);
              const rawStripped = raw.replace(/\n+$/, '');
              const rawLines = rawStripped.split('\n');
              if (rawLines.length > 0) {
                const indents = rawLines
                  .filter((ln) => ln.trim().length > 0)
                  .map((ln) => ln.length - ln.trimStart().length);
                const minIndent = indents.length > 0 ? Math.min(...indents) : 0;
                const dedented = rawLines.map((ln) =>
                  ln.length >= minIndent ? ln.slice(minIndent) : ln
                );
                fields[key] = dedented.join('\n');
              } else {
                fields[key] = '';
              }
              types[key] = 'string';
              syntax[key] = 'yaml_multiline';
              lastValidField = key;
              // Skip to list_item_close
              i = scan;
              continue;
            }

            // Fallback: collect from tokens
            const multilineParts: string[] = [];
            if (valueStr.trim().startsWith('|\n')) {
              const afterPipe = valueStr.trim().slice(2);
              const lines = afterPipe.split('\n');
              if (lines.length > 0) {
                const indents = lines
                  .filter((ln) => ln.trim().length > 0)
                  .map((ln) => ln.length - ln.trimStart().length);
                const minIndent = indents.length > 0 ? Math.min(...indents) : 0;
                const dedented = lines.map((ln) =>
                  ln.length >= minIndent ? ln.slice(minIndent) : ln
                );
                multilineParts.push(dedented.join('\n'));
              }
            }
            i++;
            let _nesting = 0;
            while (i < tokens.length) {
              const t = tokens[i];
              if (t?.type === 'list_item_close' && _nesting === 0) {
                break;
              }
              if (t?.type === 'fence') {
                // Code block - include with fences
                const lang = t.info || '';
                const fenceContent = (t.content || '').replace(/\n$/, '');
                if (lang) {
                  multilineParts.push(`\`\`\`${lang}\n${fenceContent}\n\`\`\``);
                } else {
                  multilineParts.push(`\`\`\`\n${fenceContent}\n\`\`\``);
                }
              } else if (t?.type === 'code_block') {
                // Indented code block
                multilineParts.push((t.content || '').replace(/\n$/, ''));
              } else if (t?.type === 'inline') {
                multilineParts.push(t.content || '');
              } else if (t?.type === 'ordered_list_open' || t?.type === 'bullet_list_open') {
                _nesting++;
              } else if (t?.type === 'ordered_list_close' || t?.type === 'bullet_list_close') {
                _nesting--;
              } else if (t?.type === 'list_item_open' || t?.type === 'list_item_close') {
                // skip nested list item boundaries
              }
              // Skip paragraph_open/close, list_item_open/close
              i++;
            }

            fields[key] = multilineParts.join('\n') || '';
            types[key] = 'string';
            syntax[key] = 'yaml_multiline';
            lastValidField = key;
            continue;
          }

          if (options?.rawStrings) {
            fields[key] = valueStr;
            types[key] = 'string';
            lastValidField = key;
          } else {
            const [value, typeName, rawStr, arrayRawTokens] = parseFieldValue(valueStr);
            fields[key] = value;
            types[key] = typeName === 'ref_array' ? 'array' : typeName;
            if (rawStr !== undefined) {
              rawValues[key] = rawStr;
            }
            // For arrays, record per-element raw float tokens under compound keys
            // (`key[idx]`). These never collide with real field keys (which match
            // /^[a-zA-Z_][a-zA-Z0-9_]*$/) and are read only by the JSON data-column
            // serializer to restore canonical `X.0` form; rebuild ignores them.
            if (arrayRawTokens) {
              for (let idx = 0; idx < arrayRawTokens.length; idx++) {
                const tok = arrayRawTokens[idx];
                if (tok !== undefined) {
                  rawValues[`${key}[${idx}]`] = tok;
                }
              }
            }
            lastValidField = key;

            // Track syntax for arrays
            if (typeName === 'ref_array') {
              syntax[key] = 'comma_refs';
            } else if (typeName === 'array') {
              // Detect multiline array: value spans multiple lines
              if (isMultilineArray) {
                syntax[key] = 'yaml_multiline_array';
              } else {
                syntax[key] = 'yaml_array';
              }
            }
          }

          // Detect nested sub-items: field with empty value followed by nested list
          // e.g. `- affected_files:\n  - item1\n  - item2`
          // This is a syntax error (nested_subitems) per spec rule_no_nested_subitems
          if (valueStr === '' && i + 1 < tokens.length) {
            // Look ahead for nested bullet_list_open or ordered_list_open before list_item_close
            let lookahead = i + 1;
            while (lookahead < tokens.length && tokens[lookahead]?.type === 'paragraph_close') {
              lookahead++;
            }
            const nestedType = tokens[lookahead]?.type;
            if (
              lookahead < tokens.length &&
              (nestedType === 'bullet_list_open' || nestedType === 'ordered_list_open')
            ) {
              const nestedClose = nestedType.replace('_open', '_close');
              // Skip past the nested list
              lookahead++; // skip list_open
              while (lookahead < tokens.length && tokens[lookahead]?.type !== nestedClose) {
                lookahead++;
              }
              if (lookahead < tokens.length) {
                lookahead++; // skip bullet_list_close
              }
              // Record as nested_subitems error
              nestedSubitemsErrors.push({
                key,
                line: token.map ? token.map[0] + 1 : 0,
              });
              // Remove the field we just added (it has empty value)
              delete fields[key];
              delete types[key];
              i = lookahead;
              continue;
            }
          }
        }
      } else {
        // Not a valid field - check if it looks like a field with invalid key
        const invalidMatch = firstLine.match(invalidFieldLikePattern);
        if (invalidMatch) {
          const potentialKey = (invalidMatch[1] ?? '').trim();
          if (potentialKey && !validKeyPattern.test(potentialKey)) {
            // Invalid field key (e.g. Cyrillic)
            const lineNum = token.map ? token.map[0] + 1 : 0;
            invalidItems.push({
              key: potentialKey,
              content: `- ${content}`,
              line: lineNum,
              after: lastValidField,
            });
          }
        } else if (lastValidField !== '__self' && content) {
          // Plain list item after a valid field — not a field at all.
          // Save as invalid item so it can be preserved in __comments.
          const lineNum = token.map ? token.map[0] + 1 : 0;
          invalidItems.push({
            key: '',
            content: `- ${content}`,
            line: lineNum,
            after: lastValidField,
          });
        }
      }

      i++;
      continue;
    }

    // Unknown token, skip
    i++;
  }

  return [fields, types, syntax, invalidItems, i, rawValues, nestedSubitemsErrors];
}

/**
 * Parse list items as array elements (no key: prefix)
 *
 * Used for [[field: array]] sections where list items are plain values.
 *
 * Returns: [items_list, next_index]
 */
export function parseArrayItemsFromList(tokens: Token[], startIdx: number): [unknown[], number] {
  const items: unknown[] = [];
  let i = startIdx;
  let nesting = 0;

  // Only bullet lists reach this function — ordered lists are intercepted
  // by the parser and emitted as ordered_list_in_array errors.
  while (i < tokens.length) {
    const token = tokens[i];
    if (!token) break;

    if (token.type === 'bullet_list_open' || token.type === 'ordered_list_open') {
      nesting++;
      i++;
      continue;
    }

    if (token.type === 'bullet_list_close' || token.type === 'ordered_list_close') {
      nesting--;
      if (nesting <= 0) {
        i++;
        break;
      }
      i++;
      continue;
    }

    if (token.type === 'list_item_open') {
      i++;
      continue;
    }

    if (token.type === 'list_item_close') {
      i++;
      continue;
    }

    if (token.type === 'paragraph_open') {
      i++;
      continue;
    }

    if (token.type === 'paragraph_close') {
      i++;
      continue;
    }

    if (token.type === 'inline') {
      const content = token.content?.trim() ?? '';
      const [value] = parseFieldValue(content);
      items.push(value);
      i++;
      continue;
    }

    // Unknown token, skip
    i++;
  }

  return [items, i];
}
